File: ios/Processors/EnrollmentProcessor.swift
//
//  EnrollmentProcessor.swift
//  AzifaceMobileSdk
//
//  Created by Nayara Dias on 18/04/24.
//  Copyright © 2024 Facebook. All rights reserved.
//

import FaceTecSDK
import Foundation
import UIKit

class EnrollmentProcessor: NSObject, Processor, FaceTecFaceScanProcessorDelegate,
                           URLSessionTaskDelegate
{
  public var success = false
  public var data: NSDictionary!
  public var latestNetworkRequest: URLSessionTask!
  public var viewController: AziFaceViewController!
  public var faceScanResultCallback: FaceTecFaceScanResultCallback!
  public let theme: Theme!

  init(sessionToken: String, viewController: AziFaceViewController, data: NSDictionary) {
    self.viewController = viewController
    self.data = data
    self.theme = Theme()

    super.init()

    AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: true)

    let controller = FaceTec.sdk.createSessionVC(
      faceScanProcessorDelegate: self, sessionToken: sessionToken)

    FaceTecUtilities.getTopMostViewController()?.present(
      controller, animated: true, completion: nil)
  }

  func processSessionWhileFaceTecSDKWaits(
    sessionResult: FaceTecSessionResult, faceScanResultCallback: FaceTecFaceScanResultCallback
  ) {
    self.viewController.setLatestSessionResult(sessionResult: sessionResult)
    self.faceScanResultCallback = faceScanResultCallback
    if sessionResult.status != .sessionCompletedSuccessfully {
      if latestNetworkRequest != nil {
        latestNetworkRequest.cancel()
      }

      AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false)
      faceScanResultCallback.onFaceScanResultCancel()
      print("(Aziface SDK) Status is not session completed successfully!")
      return
    }
    var parameters: [String: Any] = ["faceScan": sessionResult.faceScanBase64]
    if let auditTrailImage = sessionResult.auditTrailCompressedBase64?.first {
      parameters["auditTrailImage"] = auditTrailImage
    }
    if let lowQualityAuditTrailImage = sessionResult.lowQualityAuditTrailCompressedBase64?.first {
      parameters["lowQualityAuditTrailImage"] = lowQualityAuditTrailImage
    }
    parameters["externalDatabaseRefID"] = self.viewController.getLatestExternalDatabaseRefID()
    if let data = self.data {
      parameters["data"] = data
    }

    let dynamicRoute = DynamicRoute()
    let route = dynamicRoute.getPathUrlEnrollment3d(target: "base")

    do {
      var request = Config.makeRequest(url: route, httpMethod: "POST")
      request.httpBody = try JSONSerialization.data(withJSONObject: parameters, options: [])

      let session = URLSession(configuration: .default, delegate: self, delegateQueue: .main)
      latestNetworkRequest = session.dataTask(with: request) { [weak self] data, response, error in
        guard let self = self else { return }

        if error != nil {
          AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false)
          self.faceScanResultCallback.onFaceScanResultCancel()
          print("(Aziface SDK) An error occurred while scanning")
          return
        }

        guard let data = data else {
          AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false)
          self.faceScanResultCallback.onFaceScanResultCancel()
          print("(Aziface SDK) Data object not found!")
          return
        }

        do {
          guard
            let responseJSON = try JSONSerialization.jsonObject(with: data, options: [])
              as? [String: AnyObject]
          else {
            AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false)
            self.faceScanResultCallback.onFaceScanResultCancel()
            print("(Aziface SDK) Invalid JSON response.")
            return
          }
          guard let responseData = responseJSON["data"] as? [String: AnyObject] else {
            AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false)
            self.faceScanResultCallback.onFaceScanResultCancel()
            print("(Aziface SDK) Missing 'data' in response.")
            return
          }

          if let error = responseData["error"] as? Int, error != 0 {
            AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false)
            self.faceScanResultCallback.onFaceScanResultCancel()
            print("(Aziface SDK) Error in response")
            return
          }

          guard let scanResultBlob = responseData["scanResultBlob"] as? String,
                let wasProcessed = responseData["wasProcessed"] as? Int
          else {
            AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false)
            self.faceScanResultCallback.onFaceScanResultCancel()
            print(
              "(Aziface SDK) Missing required keys 'scanResultBlob' or 'wasProcessed' in 'data'.")
            return
          }

          if wasProcessed == 1 {
            let message = self.theme.getEnrollmentMessage(
              "successMessage", defaultMessage: "Face Scanned\n3D Liveness Proven")
            FaceTecCustomization.setOverrideResultScreenSuccessMessage(message)
            self.success = self.faceScanResultCallback.onFaceScanGoToNextStep(
              scanResultBlob: scanResultBlob)
          } else {
            AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false)
            self.faceScanResultCallback.onFaceScanResultCancel()
            print("(Aziface SDK) AziFace SDK wasn't able to process the values.")
            return
          }
        } catch {
          AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false)
          self.faceScanResultCallback.onFaceScanResultCancel()
          print("(Aziface SDK) Exception raised while attempting to parse JSON result.")
          return
        }
      }
      latestNetworkRequest?.resume()
    } catch {
      AzifaceMobileSdk.emitter.sendEvent(withName: "onCloseModal", body: false)
      faceScanResultCallback.onFaceScanResultCancel()
      print("(Aziface SDK) Exception raised while attempting HTTPS call.")
      return
    }

    DispatchQueue.main.asyncAfter(deadline: .now() + 6) {
      guard self.latestNetworkRequest.state != .completed else { return }

      let message = self.theme.getEnrollmentMessage(
        "uploadMessage", defaultMessage: "Still Uploading...")
      let uploadMessage = NSMutableAttributedString(string: message)
      self.faceScanResultCallback.onFaceScanUploadMessageOverride(
        uploadMessageOverride: uploadMessage)
    }
  }

  func urlSession(
    _ session: URLSession, task: URLSessionTask, didSendBodyData bytesSent: Int64,
    totalBytesSent: Int64, totalBytesExpectedToSend: Int64
  ) {
    let uploadProgress = Float(totalBytesSent) / Float(totalBytesExpectedToSend)
    faceScanResultCallback.onFaceScanUploadProgress(uploadedPercent: uploadProgress)
  }

  func onFaceTecSDKCompletelyDone() {
    self.viewController.onComplete()
  }

  func isSuccess() -> Bool {
    return success
  }
}

File: android/src/main/java/com/azify/processors/EnrollmentProcessor.java
package com.azify.processors;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.RequestBody;
import org.json.JSONException;
import org.json.JSONObject;
import java.io.IOException;

import com.azify.utils.DynamicRoute;
import com.azify.azifacemobilesdk.AzifaceMobileSdkModule;
import com.facebook.react.bridge.ReadableMap;
import com.facetec.sdk.*;

public class EnrollmentProcessor extends Processor implements FaceTecFaceScanProcessor {
  private boolean success = false;
  private final AzifaceMobileSdkModule module;
  private final ReadableMap data;

  public EnrollmentProcessor(String sessionToken, Context context, AzifaceMobileSdkModule module,
      ReadableMap data) {
    this.module = module;
    this.data = data;

    this.module.sendEvent("onOpen", true);
    this.module.sendEvent("onClose", false);
    this.module.sendEvent("onCancel", false);
    this.module.sendEvent("onError", false);

    FaceTecSessionActivity.createAndLaunchSession(context, EnrollmentProcessor.this, sessionToken);
  }

  public void processSessionWhileFaceTecSDKWaits(final FaceTecSessionResult sessionResult,
      final FaceTecFaceScanResultCallback faceScanResultCallback) {
    this.module.setLatestSessionResult(sessionResult);

    if (sessionResult.getStatus() != FaceTecSessionStatus.SESSION_COMPLETED_SUCCESSFULLY) {
      NetworkingHelpers.cancelPendingRequests();
      faceScanResultCallback.cancel();
      this.module.sendEvent("onOpen", false);
      this.module.sendEvent("onClose", true);
      this.module.sendEvent("onCancel", true);
      this.module.sendEvent("onError", false);
      this.module.promise.reject("Status is not session completed successfully!",
          "SessionStatusError");
      return;
    }

    DynamicRoute dynamicRoute = new DynamicRoute();
    String pathUrl = dynamicRoute.getPathUrlEnrollment3d("base");

    JSONObject parameters = new JSONObject();
    try {
      if (this.data != null) {
        parameters.put("data", new JSONObject(this.data.toHashMap()));
      }
      parameters.put("faceScan", sessionResult.getFaceScanBase64());
      parameters.put("auditTrailImage", sessionResult.getAuditTrailCompressedBase64()[0]);
      parameters.put("lowQualityAuditTrailImage", sessionResult.getLowQualityAuditTrailCompressedBase64()[0]);
      parameters.put("externalDatabaseRefID", module.getLatestExternalDatabaseRefID());
    } catch (JSONException e) {
      this.module.sendEvent("onOpen", false);
      this.module.sendEvent("onClose", true);
      this.module.sendEvent("onCancel", false);
      this.module.sendEvent("onError", true);
      module.promise.reject("Exception raised while attempting to create JSON payload for upload.",
          "JSONError");
    }

    okhttp3.Request request = new okhttp3.Request.Builder()
        .url(Config.BaseURL + pathUrl)
        .headers(Config.getHeaders("POST"))
        .post(new ProgressRequestBody(
            RequestBody.create(MediaType.parse("application/json; charset=utf-8"), parameters.toString()),
            new ProgressRequestBody.Listener() {
              @Override
              public void onUploadProgressChanged(long bytesWritten, long totalBytes) {
                final float uploadProgressPercent = ((float) bytesWritten) / ((float) totalBytes);
                faceScanResultCallback.uploadProgress(uploadProgressPercent);
              }
            }))
        .build();

    NetworkingHelpers.getApiClient().newCall(request).enqueue(new Callback() {
      @Override
      public void onResponse(@NonNull Call call, @NonNull okhttp3.Response response) throws IOException {

        String responseString = response.body().string();
        response.body().close();
        try {
          JSONObject responseJSON = new JSONObject(responseString);
          JSONObject responseJSONData = responseJSON.getJSONObject("data");
          boolean wasProcessed = responseJSONData.getBoolean("wasProcessed");
          boolean error = responseJSONData.getBoolean("error");

          if (error) {
            faceScanResultCallback.cancel();
            module.sendEvent("onOpen", false);
            module.sendEvent("onClose", true);
            module.sendEvent("onCancel", true);
            module.sendEvent("onError", true);
            module.promise.reject("An error occurred while scanning",
              "ProcessorError");
            return;
          }

          String scanResultBlob = responseJSONData.getString("scanResultBlob");

          if (wasProcessed) {
            FaceTecCustomization.overrideResultScreenSuccessMessage = AzifaceMobileSdkModule.AziTheme
              .getEnrollmentMessage("successMessage", "Face Scanned\n3D Liveness Proven");

            success = faceScanResultCallback.proceedToNextStep(scanResultBlob);
            if (success) {
              module.sendEvent("onOpen", false);
              module.sendEvent("onClose", true);
              module.sendEvent("onCancel", false);
              module.sendEvent("onError", false);
              module.promise.resolve(true);
            }
          } else {
            faceScanResultCallback.cancel();
            module.sendEvent("onOpen", false);
            module.sendEvent("onClose", true);
            module.sendEvent("onCancel", true);
            module.sendEvent("onError", true);
            module.promise.reject("AziFace SDK wasn't have to values processed!",
              "SessionNotProcessedError");
          }
        } catch (JSONException e) {
          faceScanResultCallback.cancel();
          module.sendEvent("onOpen", false);
          module.sendEvent("onClose", true);
          module.sendEvent("onCancel", true);
          module.sendEvent("onError", true);
        }
      }

      @Override
      public void onFailure(@NonNull Call call, @Nullable IOException e) {
        faceScanResultCallback.cancel();
        module.sendEvent("onOpen", false);
        module.sendEvent("onClose", true);
        module.sendEvent("onCancel", true);
        module.sendEvent("onError", true);
        module.promise.reject("Exception raised while attempting HTTPS call.", "HTTPSError");
      }
    });
  }

  public boolean isSuccess() {
    return this.success;
  }
}
